### ISSUE TITLE
-s command not working on CM4 Carrier board

### ISSUE THREAD

**Issue by Boaztheostrich** [FDH2/UxPlay#280]
I am having an issue with getting UxPlay to fill up my full screen. I can't seem to get it to fill up more than this amount of my screen:

![image](https://github.com/FDH2/UxPlay/assets/81034456/99f209ee-04aa-41b8-b2c6-fbd4f6eff7af)

I tried a variety of different resolutions with the -s command but it always displays this same framing.

If I do not specify the resolution with -s it displays a image larger than what my 800x480 screen can display:

![image](https://github.com/FDH2/UxPlay/assets/81034456/bdf297c9-058b-4108-9cc0-118467779674)

Not sure if there is a better way to approach this.

```
uxplay -bt709 -as 0 -s 800x480 -fs
```

Is the command I have been using.

**Comment by Boaztheostrich** [comment]
I'm not sure this is relevant, but if I don't use the -bt709 command I get this error:

```
GStreamer error: Internal data stream error.
*** This is a generic GStreamer error that usually means that GStreamer
*** was unable to construct a working video pipeline.

*** If you are letting the default autovideosink select the videosink,
*** GStreamer may be trying to use non-functional hardware h264 video decoding.
*** Try using option -avdec to force software decoding or use -vs <videosink>
*** to select a videosink of your choice (see "man uxplay").

*** Raspberry Pi OS with (unpatched) GStreamer-1.18.4 needs "-bt709" uxplay option
Removing connection for socket 32
raop_rtp_mirror->running is no longer true
Removing connection for socket 34
Initialized server socket(s)
```

**Comment by fduncanh** [comment]
you are using video4linux2 to access the GPU on a RPi 4B or earlier with Raspberry Pi OS bullseye.   unpatched gstreamer-1.18.4  needs the -bt709 option.  If patch gstreamer you wont need -bt709.
https://github.com/FDH2/UxPlay/wiki/Gstreamer-Video4Linux2-plugin-patches

as for the framing, I checked , that's what -vs ximagesink seems to do in -fs mode.  (a black frame) -
-vs xvimagescreen -fs does better at filling the screen, but  if I remember, your screen does not support xv extensions

**Comment by Boaztheostrich** [comment]
That makes sense, is there another videosink you would recommend I try that supports -fs and doesn't have the black box problem?

I reached out to support regarding the xv issue but haven't found a solution as of yet.

Thank you!

**Comment by fduncanh** [comment]
waylandsink supports fs
on rpi activate wayland  in raspi-config "advanced options".   (both Bullseye and Bookworm)

whether your screen supports it,  I wouldnt know.

**Comment by Boaztheostrich** [comment]
Unfortunately I have been unable to get past the login screen using wayland. I type in my password the screen turns black, and then it presents the login screen again. Not sure what is happening there.

**Comment by fduncanh** [comment]
https://www.geeky-gadgets.com/raspberry-pi-os-wayland/

If you also have a test installation of Bookworm, it might or might not be better

**Comment by Boaztheostrich** [comment]
Sounds good, I can look at installing that instead of bullseye.

Would I be able to just downgrade or upgrade gstreamer to get around the
ximagesink issue?

Ideally Waylandsink works.

Failing that do you have any idea on how to eliminate that black border?

Thanks for all your help.

**Comment by fduncanh** [comment]
You should be maintaining set of  SDHC cards with different versions of the R Pi OS operating system on them for testing.
That way you don't ruin a working system just to test.   (lite/desktop, bullseye/bookworm)

Down grading or upgrading gstreamer outside a given series ( 1.18.x, 1.20.x 1.22.x 1.24.x) is not possible on a desktop system.  Too much is linked to it.  It is possible on a lite system.   

You also should have another a more standard terminal to test on, to see whether your issues are to do with your 800x480 one

**Comment by Boaztheostrich** [comment]
Sounds good, I have spare cards I should have clarified.

1.18 seems to work fine for everything I need so o can try to get that
working on bookworm.

I have a way of testing on other terminals it’s just a pain with the
carrier board I’m using not having hdmi out. Xvimagesink worked on the
other terminal.

**Comment by fduncanh** [comment]
It will be difficult to get  gstreamer-1.18.x working on bookworm.  You will need to start with a Lite system, and not install any packages linked to gstreamer-1.22.x which if I remember correctly is what Bookworm supplies.  

* you have two issues:   (1)  ximagesink is broken on gstreamer-1.22 on R Pi (works for 1.20.x)    (and when working gives you the black border).   This could be tracked down (and should be), to allow you to use bookworm with you monitor

* your tiny monitor does not seem to support anything other than ximagesink?    Is this true for all such tiny monitors?  (If not, maybe get a new model.)  Or google for other users' experiences

I'll close this as  its not an issue uxplay can solve.

**Comment by Boaztheostrich** [comment]
Thanks for the response, I’ll have to look at the bookworm option a bit
more and go from there. Thanks for the help!

**Comment by fduncanh** [comment]
I took a look at the ximagesink bug.   It dates to March 2022 If you rebuild libgstximagesink.so on bookworm after replacing ximagesink.c and ximagesink.h  with the older version it works.

You rebuild gstreamer1.0-plugins-base with debuild, replace the two files, run ninja, then strip libgstximagesink.so and copy it to the /lib/.../gstreamer-1.0/ directory.   (Same procedure you did earlier to fix  the videoflip bug.)

 get the two replacement files here: https://gitlab.freedesktop.org/gstreamer/gstreamer/-/tree/2fd28195caf4502c7fc10cb3d63974c543354180/subprojects/gst-plugins-base/sys/ximage

I reported the issue to gstreamer: https://gitlab.freedesktop.org/gstreamer/gstreamer/-/issues/3470

**Comment by fduncanh** [comment]
I located the precise point where the ximagesink changes in 1.22 crash X11 on R Pi.


See https://gitlab.freedesktop.org/gstreamer/gstreamer/-/issues/3470

**Comment by Boaztheostrich** [comment]
That’s awesome! Thanks for taking the time to deep dive into this. I looked
through your comments on the link and previous response and I can try to
follow those steps today or tmw if I have time.

I’ll make sure to report back if I can get it working on my side, I would
much prefer to use bookworm if possible.

**Comment by fduncanh** [comment]
Found the problem.

```
#ifdef HAVE_XI2
    if (ximagesink->xcontext->use_xi2) {
      XIEventMask mask_data;
      unsigned char mask[2];

      gst_x_image_sink_xwindow_select_touch_events (ximagesink, xwindow);

      XISetMask (mask, XI_HierarchyChanged);
      mask_data.deviceid = XIAllDevices;
      mask_data.mask_len = sizeof (mask);
      mask_data.mask = mask;

      /* register for HierarchyChanged events to see device changes */
      XISelectEvents (ximagesink->xcontext->disp, xwindow->win, &mask_data, 1);
    }
#endif

```
line  578 should be changed to initialize mask[2]:

```
      unsigned char mask[2] = {0x0, 0x0};
```

**Comment by Boaztheostrich** [comment]
Perfect thank you!!

**Comment by Boaztheostrich** [comment]
@fduncanh do you have any advice on installing the version of gstreamer that has the fixes on RPI? I've been having some difficulty figuring it out. GStreamer 1.22.12 old-stable bug-fix is the correct version but just doing the basic:

```
sudo apt install gstreamer1.0=1.22.12
```

Didn't work.

I tried following the instructions here: https://gstreamer.freedesktop.org/documentation/installing/building-from-source-using-meson.html?gi-language=c

But I was unable to get past the cloning step:

```
boazburnett@raspberrypi:~ $ git clone https://gitlab.freedesktop.org/gstreamer/gstreamer.git
Cloning into 'gstreamer'...
remote: Enumerating objects: 913398, done.
remote: Counting objects: 100% (2023/2023), done.
remote: Compressing objects: 100% (1210/1210), done.
error: RPC failed; curl 56 Recv failure: Connection reset by peer
error: 20890 bytes of body are still expected
fetch-pack: unexpected disconnect while reading sideband packet
fatal: early EOF
fatal: fetch-pack: invalid index-pack output
```

Trying:

```
git clone https://github.com/GStreamer/gstreamer.git
```

Does work and I was able to get to the 
```
[(11/12 @ 98%) INSTALL: cargo-c]
```
step but have not been able to make it past that.

I've been following these instructions: https://gstreamer.freedesktop.org/documentation/installing/building-from-source-using-cerbero.html?gi-language=c

Just swapping the gitlab stuff for github.

Thanks!
Boaz

**Comment by fduncanh** [comment]
The ximagesink fix is in gstreamer-1.22.12 

The reference is:

https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/6655

Unfortunately, Raspbery Pi OS will not  update their gstreamer-1.22.0 package unless this issue is brought to their attention.

If you just need the fix for yourself, the easiest thing is to just patch libgstximagesink.so. with the fix.

If you are planing a device you want to distribute, its not easy to replace the distribution 's gstreamer by compiling it yourself.    The best would be to get R Pi OS to include the fix in an update.  The above reference would be useful in getting them to do it.

**Comment by Boaztheostrich** [comment]
I installed Gstreamer 1.22.12 on bookwork on my cm4. It took forever but I eventually got it up and running.

I still encountered the black box around the screen mirroring though:

![IMG_4984](https://github.com/FDH2/UxPlay/assets/81034456/75f05f9e-d2aa-4401-b867-c7c44dc626e7)
![IMG_4985](https://github.com/FDH2/UxPlay/assets/81034456/32ecd937-5961-4657-a241-2610e349a028)

Not sure if this is related but upon reboot it changed my desktop enviroment to a different flavor of linux: 
![IMG_4987](https://github.com/FDH2/UxPlay/assets/81034456/02299892-3014-4af2-8760-0d37552e7571)

I may screw around with it some more, but I'd love to hear your thoughts, I can just try to do the patch instead of installing the full 1.22.12

**Comment by fduncanh** [comment]
The "black box" is not affected by the fix for ximagesink I got added to gstreamer-1.22-12 and later.
 
Its a "feature".    Maybe by adding  ... ! videoscale !... to the video pipeline it can be adjusted: see below.

https://gstreamer.freedesktop.org/documentation/videoconvertscale/videoscale.html?gi-language=c

The pipeline is created in renderers/video_renderer_gstreamer.c  starting at line 150

https://gstreamer.freedesktop.org/documentation/tutorials/basic/platform-specific-elements.html?gi-language=c
* ximagesink

A standard RGB only X-based video sink. It implements the VideoOverlay interface, so the video window can be re-parented (embedded inside other windows). It does not support scaling or color formats other than RGB; it has to be performed by different means (using the videoscale element, for example).

* xvimagesink

An X-based video sink, using the [X Video Extension](http://en.wikipedia.org/wiki/X_video_extension) (Xv). It implements the VideoOverlay interface, so the video window can be re-parented (embedded inside other windows). It can perform scaling efficiently, on the GPU. It is only available if the hardware and corresponding drivers support the Xv extension.
